### PR TITLE
 Implement triangular array creation routines

### PR DESCRIPTION
**Pull request by tsurumeso**
This PR implements cupy.tri, cupy.tril, and cupy.triu.

### CODE REVIEW

File: cupy/creation/matrix.py [hunk 1]
```diff
@@ -51,13 +51,81 @@ def diagflat(v, k=0):
...
+
+    Returns:
+        cupy.ndarray: An array with ones at and below the given diagonal.
+
+    .. seealso:: :func:`numpy.tri`
+
+    """
+    if M is None:
+        M = N
+    out = cupy.zeros((N, M), dtype=dtype)
```
Comment (okuta) [inline cupy/creation/matrix.py]: How about following code?
```
out = cupy.empty((N, M), dtype=dtype)
return cupy.ElementwiseKernel(
        'int32 m, int32 k',
        'T out',
        '''
        int row = i % m;
        int col = i / m;
        out = (row <= col + k);
        ''',
       'tri',
    )(M, k, out)

File: cupy/creation/matrix.py [hunk 2]
```diff
@@ -51,13 +51,81 @@ def diagflat(v, k=0):
...
+            diagonal, a positive value is above it, and a negative value is
+            below.
+
+    Returns:
+        cupy.ndarray: An upper triangle of an array.
+
+    .. seealso:: :func:`numpy.triu`
+
+    """
+    if not isinstance(m, cupy.ndarray):
```
Comment (okuta) [inline cupy/creation/matrix.py]: Is is not needed. You can always use `m = cupy.asarray(m)`.

File: cupy/creation/matrix.py [hunk 3]
```diff
@@ -51,13 +51,81 @@ def diagflat(v, k=0):
...
+        m = cupy.asarray(m)
+    mask = tri(*m.shape[-2:], k=k-1, dtype=bool)
 
-# TODO(okuta): Implement triu
+    return cupy.where(mask, cupy.zeros(1, m.dtype), m)
```
Comment (okuta) [inline cupy/creation/matrix.py]: How about `return cupy.where(mask, 0, m)` ?

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/creation/matrix.py (hunk 3 above)
Comment (tsurumeso) [reply cupy/creation/matrix.py]: It caused `dtype` mismatch between `numpy` and `cupy` on the tests. I think no changes needed.

Comment (okuta) [reply cupy/creation/matrix.py]: OK, I understand.
How about using `m.dtype.type(0)` ?

**Comment by tsurumeso** [comment]
Thank you for your review. I fixed.

**Comment by okuta** [comment]
Thanks!
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**